core/domain/stats_jobs_one_off.py

```diff
@@ -18,14 +18,541 @@
+                    state_hit_event_dicts_sessioned, key=lambda e: e[
+                        'created_on'])
+                # All state hit events in the above list are ordered by their
+                # timestamp. This implies that the last state hit event is the
+                # state at which the learner quit the exploration (This is only
+                # True if the last state hit event is not the terminal state).
+                # So, if the last state is not the terminal state, we increment
+                # num_completions for all states except the last state.
+                # Otherwise, we increment num_completions for all the states.
+                for state_hit_event in state_hit_event_dicts_sessioned[:-1]:
+                    state_stats_mapping[state_hit_event[
+                        'state_name']].num_completions_v1 += 1
+                if session_id in completed_session_ids:
+                    state_stats_mapping[state_hit_event_dicts_sessioned[-1][
+                        'state_name']].num_completions_v1 += 1
+
+            # Compute total_answers_count and useful_feedback_count.
+            for state_name in state_answer_counts_for_this_version:
+                state_stats_mapping[state_name].total_answers_count_v1 += (
+                    state_answer_counts_for_this_version[state_name][
+                        'total_answers_count'])
+                state_stats_mapping[state_name].useful_feedback_count_v1 += (
+                    state_answer_counts_for_this_version[state_name][
+                        'useful_feedback_count'])
+
+            # Now that the stats for the exploration's version are computed,
+            # calculate num_actual_starts. To compute this, we take the max of
+            # the first hit counts of all states that the initial state leads
+            # to. The max is taken instead of sum because the sum of
+            # first_hit_counts will sometimes end up counting a single learner
+            # twice. The max of both first hit counts ensures that we are
+            # counting only unique learners who traversed past the initial
+            # state.
+            init_state = versioned_exploration.states[
+                versioned_exploration.init_state_name]
+            max_first_hit_from_init_state = max([
+                state_stats_mapping[
+                    answer_group.outcome.dest].first_hit_count_v1
+                for answer_group in init_state.interaction.answer_groups if (
+                    answer_group.outcome.dest != (
+                        versioned_exploration.init_state_name))] or [0])
+            # If the exploration has only one state, then the number of actual
+            # starts will be equal to the number of starts.
+            if len(versioned_exploration.states) == 1:
+                num_actual_starts = num_starts
+            else:
+                num_actual_starts = max_first_hit_from_init_state
```

Comment:
You might as well put the above calculation in this branch so that it's not done unnecessarily?

Reply:
Done.